### PR TITLE
ADR: use separate repo

### PR DESCRIPTION
**Pull request by lucyb**
This documents our use of a separate repo for the Dockerfile and other supporting files needed for the research-template dev container configuration.

Closes opensafely-core/codespaces-initiative#63

### CODE REVIEW

File: docs/adr/0001-create-research-template-docker-repo.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+Date: 2024-06-04
+
+## Status
+
+Accepted
+
+## Context
+
+We are using [dev containers](https://code.visualstudio.com/docs/devcontainers/create-dev-container) to provide a configuration for Codespaces that gives researchers a useful development environment to use while working on their OpenSAFELY projects. We have created a custom Docker image for this dev container environment. This custom image requires a Dockerfile and other supporting files to be created.
```
Comment (iaindillingham) [inline docs/adr/0001-create-research-template-docker-repo.md:11]: Nygard states that this section should describe "the forces at play". I don't think the Dockerfile is a force. We could copy it from repository to repository, just as we copy `.devcontainer/devcontainer.json`. However, I think minimising redundancy is a force. Furthermore, as this is the "root" ADR, I think we should expand on why we have decided to use dev containers, Docker images, and Codespaces. What about:

> We know that researchers often struggle to configure their development environments, which degrades their experience of working with OpenSAFELY. In response, we decided to configure a standard development environment using a custom dev container and Docker image, which researchers will interact with in Codespaces.
>
> The code for each OpenSAFELY project lives within a separate repository within the opensafely organisation. Whilst some configuration for the standard development environment _must_ live alongside the code -- because Codespaces are started from repositories -- this results in redundancy. We want to minimise redundancy, and hence to minimise the cost of maintaining the standard development environment.

File: docs/adr/0001-create-research-template-docker-repo.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+## Decision
+
+We will create a separate GitHub repo to store the Dockerfile and other supporting files. Specifically, we will store the following in the research-template-docker repo:
+
+* The Dockerfile used to generate the dev containers Docker image
+* The tooling and GitHub actions to build the Docker image and to run the tests
+* Some of the supporting files for the research-template (e.g documentation and ADRs).
```
Comment (iaindillingham) [inline docs/adr/0001-create-research-template-docker-repo.md:19]: But _why_? For example:

> We will store the Dockerfile in research-template-docker because we do not want researchers to modify it and because we want to version it.

File: docs/adr/0001-create-research-template-docker-repo.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+The research-template repo will only contain the code and configuration to be copied into a researcher’s repo.
+
+## Consequences
+
+By storing the supporting infrastructure in a separate repo we will be creating a good separation of concerns between the research code that will be used to help researchers get started on OpenSAFELY and the supporting infrastructure. The researchers who base their work on the research-template repo will have a copy of the things they need to get started and won't have any of the additional supporting infrastructure. They will not even need to be aware that the supporting infrastructure exists. That supporting infrastructure will be controlled and updated by developers working in the opensafely-core GitHub organisation. Storing the repo in a separate organisation also allows us to apply different security settings to the repo and to more easily setup Dependabot, etc.
```
Comment (iaindillingham) [inline docs/adr/0001-create-research-template-docker-repo.md:25]: Nygard states that this section should describe the context after the decision has been implemented. There are reasons why locating the configuration in two repositories is a good decision, a neutral decision, and a bad decision. I can't see any reasons why it is a neutral or a bad decision, here. Surely this isn't the best of all possible worlds?

If we had the choice, then we probably wouldn't locate _any_ configuration for the standard development environment alongside the OpenSAFELY project code. However, we don't have the choice.

**Review by iaindillingham — COMMENTED** [review]
As this is the project's -- and my! -- first ADR, I've left several comments. I'd be happy to talk more about them; rework the PR; do something else 🙂

File: docs/adr/0001-create-research-template-docker-repo.md (hunk 1 above)
Comment (lucyb) [reply docs/adr/0001-create-research-template-docker-repo.md]: That's a good point. I think the force in this case was the need to pre-build the Docker image. I've updated the text to make that clearer.

We were already using dev containers and Codespaces before we decided to create this repo and even before we started this initiative. It's probably a decision we should record, but I don't think this ADR is the right place.

File: docs/adr/0001-create-research-template-docker-repo.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# 1. Create the research-template-docker repo
```
[the remark this answers is not in the document]
Comment (lucyb) [reply docs/adr/0001-create-research-template-docker-repo.md]: Good spot. I've renamed. I'm not sure if it's a noun phrase or not though.

File: docs/adr/0001-create-research-template-docker-repo.md (hunk 2 above)
Comment (lucyb) [reply docs/adr/0001-create-research-template-docker-repo.md]: I've tried to address that in the consequences section, which I've reworded in my most recent commit. I've tried to keep the decision section short and to just state the decision without the reasoning.

File: docs/adr/0001-create-research-template-docker-repo.md (hunk 3 above)
Comment (lucyb) [reply docs/adr/0001-create-research-template-docker-repo.md]: I've tried to reword this to make it a bit clearer and to address your other comment. Please let me know if that's helped.

File: docs/adr/0001-create-research-template-docker-repo.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (iaindillingham) [reply docs/adr/0001-create-research-template-docker-repo.md]: If you can put "the", "a", or "an" in front of it, then it's a noun phrase.

[the remark this answers is not in the document]
Comment (lucyb) [reply docs/adr/0001-create-research-template-docker-repo.md]: Excellent. In that case, I think I've done that.

File: docs/adr/0001-create-research-template-docker-repo.md (hunk 1 above)
Comment (iaindillingham) [reply docs/adr/0001-create-research-template-docker-repo.md]: Let me check my understanding...

The architecture decision is "Store `x` in opensafely/research-template and `y` in opensafely-core/research-template-docker", where `x` is something like "dev container configuration" and `y` is something like "Docker image configuration".

We made the architecture decision because it means we can pre-build the Docker image. But why do we want to pre-build the Docker image? The answers to this question are the "forces at play". We want to pre-build the Docker image:

* so the researcher doesn't have to
* to reduce start up time

Comment (lucyb) [reply docs/adr/0001-create-research-template-docker-repo.md]: Your understanding is correct. I've added some additional context to explain why we're pre-building the image.
